File: back/src/main/java/com/back/domain/ticket/listener/PaymentCompletedEventHandler.java
package com.back.domain.ticket.listener;

import com.back.domain.ticket.event.PaymentCompletedEvent;
import com.back.domain.waiting.service.WaitingQueueManager;
import com.back.domain.waiting.service.WaitingQueueService;
import lombok.RequiredArgsConstructor;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

@Component
@RequiredArgsConstructor
public class PaymentCompletedEventHandler {

    private final WaitingQueueManager waitingQueueManager;
    private final WaitingQueueService waitingQueueService;

    @TransactionalEventListener(
            phase = TransactionPhase.AFTER_COMMIT
    )
    public void handlePaymentCompleted(
            PaymentCompletedEvent event
    ) {
        waitingQueueManager.removeActiveUser(
                event.scheduleId(),
                event.userId()
        );

        waitingQueueService.allowEntry(
                event.concertId(),
                event.scheduleId()
        );
    }
}

File: back/src/main/kotlin/com/back/domain/ticket/listener/PaymentCompletedEventHandler.kt
package com.back.domain.ticket.listener

import com.back.domain.ticket.event.PaymentCompletedEvent
import com.back.domain.waiting.service.WaitingQueueManager
import com.back.domain.waiting.service.WaitingQueueService
import org.springframework.stereotype.Component
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

@Component
class PaymentCompletedEventHandler(
    private val waitingQueueManager: WaitingQueueManager,
    private val waitingQueueService: WaitingQueueService
) {

    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT, fallbackExecution = true)
    fun handlePaymentCompleted(event: PaymentCompletedEvent) {
        waitingQueueManager.removeActiveUser(event.scheduleId, event.userId)
        waitingQueueService.allowEntry(event.concertId, event.scheduleId)
    }
}
